packages/common/src/store/ui/modals/createModal.ts

```diff
@@ -0,0 +1,97 @@
+import { useCallback, useRef } from 'react'
+
+import { createSlice, PayloadAction } from '@reduxjs/toolkit'
+import { useDispatch, useSelector } from 'react-redux'
+
+import { CommonState } from 'store/index'
+
+export type BaseModalState = {
+  isOpen: boolean
+}
+
+/**
+ * Creates the necessary actions/reducers/selectors for a modal,
+ * and returns the reducer to be wired to the store
+ * and hook to be used in the application.
+ */
+export const createModal = <T>({
+  reducerPath,
+  initialState,
+  sliceSelector
+}: {
+  reducerPath: string
+  initialState: T & BaseModalState
+  sliceSelector?: (state: CommonState) => Record<string, any>
+}) => {
+  const slice = createSlice({
+    name: `modals/${reducerPath}`,
+    initialState,
+    reducers: {
+      setState: (_, action: PayloadAction<T & BaseModalState>) => {
+        return action.payload
+      }
+    }
+  })
+
+  const selector = (state: CommonState) => {
+    let baseState: (T & BaseModalState) | undefined
+    if (sliceSelector) {
+      baseState = sliceSelector(state)[reducerPath]
+    } else {
+      baseState = state[reducerPath]
+    }
+    if (!baseState) {
+      throw new Error(
+        `State for ${reducerPath} is undefined - did you forget to register the reducer in @audius/common/src/modals/reducers.ts?`
+      )
+    }
+    return baseState
+  }
+
+  // Need to explicitly retype this because TS got confused
+  const setState = slice.actions.setState as (
+    state: T & BaseModalState
+  ) => PayloadAction<T & BaseModalState>
+
+  /**
+   * A hook that returns the state of the modal,
+   * a setter callback that opens the modal,
+   * and a close callback that clears the state and closes it.
+   * @returns [state, open, close] state of the modal, callback to open the modal, callback to close the modal
+   */
+  const useModal = () => {
+    // Use a ref to prevent flickers on close when state is cleared.
```

Comment:
Would love to understand more what the problem is here. I feel like refs are last resorts and I'm wondering if we have a state management issue if we're seeing flickering. Generally the pattern of `setIsOpen(false) => modal disappears` should be a straightforward pattern?

Reply:
Problem: How do we cleanup stale data out of the store when modals are closed, _without_ having that modal change appearance during the "close" animation?

Perhaps the problem statement is flawed, as we force a reset of the state on open anyway and the cost of a larger redux store is probably not too huge, but personally I'm a fan of cleaning up on close

Reply:
Ah I think I see. Could use an effect cleanup function?

Reply:
We'd have to have something fire on animation complete I think, which might be how `useDrawer` works. I'll look into it!

Reply:
Done!